langchain/text_splitter.py

```diff
@@ -1003,6 +1003,126 @@ def get_separators_for_language(language: Language) -> List[str]:
             )
 
 
+class BalancedRecursiveCharacterTextSplitter(TextSplitter):
```

Comment:
Thanks for adding! Can this logic be added to the existing `RecursiveCharacterTextSplitter`?

I'm slightly concerned about addition of a large amt of new code w/ very similar functionality.

Reply:
The submitted BalancedRecursiveCharacterTextSplitter can be quite a bit slower, so I'm not sure if I want to force it on anyone. Hence why I didn't just rewrite RecursiveCharacterTextSplitter.

However, RecursiveCharacterTextSplitter should, in it's current state, be viewed as broken. Unless the total token count of a document is evenly divisible by the chunk_size param, you will always get a final chunk that is smaller than chunk_size. And in many cases it's too small to be useful, so it's essentially chopping the end off of most documents. 

I believe the correct course of action is to rewrite the RecursiveCharacterTextSplitter but with a better implementation than mine. But that's beyond my skill level dealing with recursion and efficiency. In the mean time, I'm submitting my changes in the hope someone more talented than me can rewrite it. (Also, so I don't have to maintain a fork of Langchain for my personal use.)